### PR TITLE
Competitive Programming: Binary Search Documentation

### LINKED ISSUE #949: Competitive Programming: Binary Search Problems
<hr>

## Description 📜

Adding questions related to binary search in documentation format
<hr>

## Domain of Contribution 📊

- [x] Competitive Programming

<hr>

### PR DESCRIPTION
**Pull request by ayushijindal**
<hr>

## Format: Documentation
Made the documentation on the binary search algorithm.

## AIM
Added code to explain the algorithm and made documentation in the .md format

## Acknowledgment
To my mentor @atarax665 helping me making commit and PR for this issue, a big shoutout to him 🥳

do let me know if there are still any more changes that need to be done👍
👍

Fixes #949

<hr>

## Type of change 📝

- [ ] Audio (Should be in mp3 format Includes speech clarity, Concise ,Low distortion)
- [ ] Video (Animations, screen-recordings, presentations and regular explanatory films are all possibilties etc)
- [x] Documentation (Content Creation in the form of codes or tutorials)
- [ ] Other (If you choose other, Please mention changes below) 

<hr>

## Domain of Contribution 📊

- [x] Competitive #Programming

<hr>

## Checklist ✅

- [x] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.
- [x] I have performed a self-review of my own code or work.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generates no new warnings.
- [x] I'm GWOC'21 contributor

<hr>

## Screenshots / Gif (Optional) 📸

<hr>

### CODE REVIEW

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+### Approach: 
+1. Compare x with the middle element of array A.
+2. If x matches with the middle element, we return the mid index.
+3. Else If x is greater than the mid element, then x can only lie in the right half subarray after the mid element. So we recur for the right half.
+4. Else (x is smaller) recur for the left half.
+#### Note: To apply binary search array must be sorted.
+
+#### Code in C++:
+
+```c
```
Comment (atarax665) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md:31]: Add pseudo-code of binary search.

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+## Steps To Implement Binary Search
+Let us assume there is an array A and an element x.
+We have to search throughout the array and we need to find element x inside the array.
+If not present print -1 or print the position of the array.
+
+
+---
+<p align="center">
+  <img width="460" height="300" src="https://miro.medium.com/max/247/1*fMQ2cLdIEr3DoHitYyiSsg.png">
```
Comment (atarax665) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md:17]: Can you please use a clearer image? And please give image credit if you are using an image from a different source.

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+   int size=sizeOf(arr)/sizeOf(arr[0]);
+   
+   // Calling of binary_search function
+   int index = binary_search(arr, size, target);
+   cout<<index<<endl;
+   return 0;
+}
+```
+
+## Binary Search Problems
```
Comment (atarax665) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md:78]:
```suggestion
## Some common binary search problems
```

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+### Question: 
+### Find the index of first 1 in a sorted array of 0’s and 1’s if 1 is not present in the array print -1
```
Comment (atarax665) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md:80]:
```suggestion
### Find the index of first 1 in a sorted array of 0’s and 1’s if 1 is not present in the array then print -1
```

File: Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        cout<<"Not found"<<endl;
+    
+
+}
+```
+
+## End Note
+Hope this documentation provided you some basic idea about binary search algorithm and how to solve questions using this approach.
+Binary search runs in logarithmic time in the worst case, making O(log n) comparisons, where n is the number of elements in the array.Binary search is faster than linear search except for small arrays. However, the array must be sorted first to be able to apply binary search. There are specialized data structures designed for fast searching, such as hash tables, that can be searched more efficiently than binary search. However, binary search can be used to solve a wider range of problems, such as finding the next-smallest or next-largest element in the array relative to the target even if it is absent from the array.
+
```
Comment (atarax665) [inline Competitive_Programming/C++/Searching & Sorting/Binary Search/binary_search_documentation.md:238]: Remove these extra blank lines.

**Comment by ayushijindal** [comment]
Ok I will do the changes ASAP

**Comment by ayushijindal** [comment]
@atarax665 I have made the required modifications. If it still needs further changes then kindly inform me  :smiley:

**Comment by atarax665** [comment]
Please do not merge the main branch with mybranch as you are making a contribution to the Competitive_Programming branch which is not synced with the main branch.

**Comment by ayushijindal** [comment]
@atarax665  Can you assist me in making this issue resolved?

**Pull request closed without merging**